File: skills/agents-orchestrator/scripts/hook_runtime.py
#!/usr/bin/env python3
"""Minimal Runtime surface invoked by the deployed Claude Hooks."""

import argparse
import json
import os


OWNER_LEASE_SECONDS = 15 * 60
IDENTITY_ENVIRONMENT = {
    "root_id": "ROOT_ID",
    "task_id": "TASK_ID",
    "attempt_id": "ATTEMPT_ID",
    "actor_token": "ACTOR_TOKEN",
}
CANONICAL_PREFIX = "AGENTS_ORCHESTRATOR_"
LEGACY_PREFIX = "AGENT_SWARM_"


def _environment_value(suffix):
    primary_name = CANONICAL_PREFIX + suffix
    legacy_name = LEGACY_PREFIX + suffix
    primary = os.environ.get(primary_name, "").strip() or None
    legacy = os.environ.get(legacy_name, "").strip() or None
    if primary and legacy and primary != legacy:
        raise ValueError(
            "conflicting orchestration environment: %s does not match %s"
            % (primary_name, legacy_name)
        )
    return primary or legacy


def _prepare_runtime_home():
    runtime_home = _environment_value("HOME")
    if runtime_home:
        os.environ[CANONICAL_PREFIX + "HOME"] = runtime_home
        os.environ[LEGACY_PREFIX + "HOME"] = runtime_home


_prepare_runtime_home()

# The deployed Hook bundle intentionally contains only this module plus the
# existing store/manager.  Keep its small compatibility bootstrap self-contained.
import hook_manager  # noqa: E402
import state_store  # noqa: E402


def _verify_owner(run, token):
    if not state_store.token_matches(token, run.get("owner_token_hash")):
        raise RuntimeError("invalid owner token")


def heartbeat(root_id, task_id, attempt_id, actor_token):
    with state_store.transaction() as con:
        run = state_store.get_run(root_id, con)
        task = state_store.get_task(task_id, con)
        attempt = state_store.get_attempt(attempt_id, con)
        if not all((run, task, attempt)):
            raise RuntimeError("invalid heartbeat binding")
        if not (
            task["root_id"] == root_id
            and attempt["root_id"] == root_id
            and attempt["task_id"] == task_id
        ):
            raise RuntimeError("invalid heartbeat binding")
        if not state_store.token_matches(actor_token, attempt["actor_token_hash"]):
            raise RuntimeError("invalid actor token")
        current = state_store.get_current_attempt(task_id, con)
        if current is None or current["attempt_id"] != attempt_id:
            return {"accepted": False, "stale_attempt": True}
        if run["status"] != "running" or attempt["state"] in {"done", "failed", "cancelled"}:
            return {"accepted": False, "terminal": True}
        timestamp = state_store.now()
        con.execute("UPDATE attempts SET heartbeat_at=? WHERE attempt_id=?", (timestamp, attempt_id))
        lease_expires_at = run.get("lease_expires_at")
        if task_id == run["root_task_id"]:
            _verify_owner(run, actor_token)
            if lease_expires_at and lease_expires_at < timestamp:
                raise RuntimeError("root owner lease expired; recover the run")
            lease_expires_at = timestamp + OWNER_LEASE_SECONDS
            con.execute(
                "UPDATE runs SET lease_expires_at=?, updated_at=? WHERE root_id=?",
                (lease_expires_at, timestamp, root_id),
            )
        return {
            "accepted": True,
            "heartbeat_at": timestamp,
            "lease_expires_at": lease_expires_at,
        }


def observe_session_end(root_id, task_id, attempt_id, actor_token):
    """Record SessionEnd without changing Attempt or Task lifecycle state."""
    with state_store.transaction() as con:
        run = state_store.get_run(root_id, con)
        task = state_store.get_task(task_id, con)
        attempt = state_store.get_attempt(attempt_id, con)
        if not all((run, task, attempt)):
            raise RuntimeError("invalid SessionEnd binding")
        if not (
            task["root_id"] == root_id
            and attempt["task_id"] == task_id
            and state_store.token_matches(actor_token, attempt["actor_token_hash"])
        ):
            raise RuntimeError("invalid SessionEnd identity")
        state_store.append_event(
            con, root_id, "SessionEndObserved", {"attempt_state": attempt["state"]},
            task_id=task_id, attempt_id=attempt_id,
        )
        return {"observed": True, "attempt_id": attempt_id, "state": attempt["state"]}


def _authorize_read(root_id, actor_token):
    state_store.initialize_schema()
    run = state_store.get_run(root_id)
    if run is None:
        raise ValueError("run not found")
    if not any(
        state_store.token_matches(actor_token, attempt["actor_token_hash"])
        for attempt in state_store.list_attempts(root_id)
    ):
        raise ValueError("invalid actor token")
    return run


def inspect_current(root_id, task_id, actor_token):
    run = _authorize_read(root_id, actor_token)
    task = state_store.get_task(task_id)
    if task is None or task["root_id"] != root_id:
        raise ValueError("current task does not belong to the authorized run")
    attempt = state_store.get_current_attempt(task_id)
    if attempt is None or attempt["root_id"] != root_id or attempt["task_id"] != task_id:
        raise ValueError("current attempt binding is invalid")
    launch = state_store.get_current_launch(attempt["attempt_id"])
    session = state_store.get_session_for_launch(launch["launch_id"]) if launch else None
    return {"run": run, "task": task, "attempt": attempt, "launch": launch, "session": session}


def _refresh_run_hooks(root_id, cwd):
    if state_store.get_run(root_id) is not None:
        hook_manager.ensure_project_hooks(cwd, root_id=root_id)


def _identity_from_environment():
    suffixes = tuple(IDENTITY_ENVIRONMENT.values())
    for prefix in (CANONICAL_PREFIX, LEGACY_PREFIX):
        present = [
            suffix
            for suffix in suffixes
            if os.environ.get(prefix + suffix, "").strip()
        ]
        if present and len(present) != len(suffixes):
            missing = [
                prefix + suffix for suffix in suffixes if suffix not in present
            ]
            raise ValueError(
                "partial orchestration identity: missing %s" % ", ".join(missing)
            )
    values = {
        key: _environment_value(suffix)
        for key, suffix in IDENTITY_ENVIRONMENT.items()
    }
    missing = [
        CANONICAL_PREFIX + IDENTITY_ENVIRONMENT[key]
        for key, current in values.items()
        if not current
    ]
    if missing:
        raise ValueError(
            "missing orchestration identity: %s" % ", ".join(missing)
        )
    for key, suffix in IDENTITY_ENVIRONMENT.items():
        os.environ[CANONICAL_PREFIX + suffix] = values[key]
        os.environ[LEGACY_PREFIX + suffix] = values[key]
    for key in ("task_id", "attempt_id"):
        try:
            values[key] = int(values[key])
        except (TypeError, ValueError) as exc:
            raise ValueError(
                "%s must be an integer"
                % (CANONICAL_PREFIX + IDENTITY_ENVIRONMENT[key])
            ) from exc
    return values


def main(argv=None):
    parser = argparse.ArgumentParser(prog="hook_runtime.py")
    parser.add_argument("command", choices=("heartbeat", "inspect-current", "session-end"))
    args = parser.parse_args(argv)
    try:
        values = _identity_from_environment()
        if args.command == "heartbeat":
            _refresh_run_hooks(values["root_id"], os.getcwd())
            result = heartbeat(**values)
        elif args.command == "inspect-current":
            result = inspect_current(
                values["root_id"], values["task_id"], values["actor_token"]
            )
        else:
            result = observe_session_end(**values)
        print(json.dumps(result, ensure_ascii=False))
        return 0
    except (ValueError, RuntimeError) as exc:
        parser.error(str(exc))


if __name__ == "__main__":
    main()

File: skills/agents-orchestrator/scripts/hook_runtime.ts
#!/usr/bin/env bun
import * as compatEnv from "./compat_env.ts";
import * as hookManager from "./hook_manager.ts";
import * as stateStore from "./state_store.ts";
import { RuntimeError, ValueError, type RuntimeRecord } from "./runtime_types.ts";

export const OWNER_LEASE_SECONDS = 15 * 60;

function verifyOwner(run: RuntimeRecord, token: string): void {
  if (!stateStore.tokenMatches(token, run.owner_token_hash)) throw new RuntimeError("invalid owner token");
}

export function heartbeat(rootId: string, taskId: number, attemptId: number, actorToken: string): RuntimeRecord {
  return stateStore.transaction((connection) => {
    const run = stateStore.getRun(rootId, connection);
    const task = stateStore.getTask(taskId, connection);
    const attempt = stateStore.getAttempt(attemptId, connection);
    if (!run || !task || !attempt || task.root_id !== rootId || attempt.root_id !== rootId || attempt.task_id !== taskId) {
      throw new RuntimeError("invalid heartbeat binding");
    }
    if (!stateStore.tokenMatches(actorToken, attempt.actor_token_hash)) throw new RuntimeError("invalid actor token");
    const current = stateStore.getCurrentAttempt(taskId, connection);
    if (!current || current.attempt_id !== attemptId) return { accepted: false, stale_attempt: true };
    if (run.status !== "running" || new Set(["done", "failed", "cancelled"]).has(String(attempt.state))) {
      return { accepted: false, terminal: true };
    }
    const timestamp = stateStore.now();
    connection.execute("UPDATE attempts SET heartbeat_at=? WHERE attempt_id=?", [timestamp, attemptId]);
    let leaseExpiresAt = run.lease_expires_at;
    if (taskId === run.root_task_id) {
      verifyOwner(run, actorToken);
      if (leaseExpiresAt && Number(leaseExpiresAt) < timestamp) throw new RuntimeError("root owner lease expired; recover the run");
      leaseExpiresAt = timestamp + OWNER_LEASE_SECONDS;
      connection.execute("UPDATE runs SET lease_expires_at=?, updated_at=? WHERE root_id=?", [leaseExpiresAt, timestamp, rootId]);
    }
    return { accepted: true, heartbeat_at: timestamp, lease_expires_at: leaseExpiresAt };
  });
}

export function observeSessionEnd(rootId: string, taskId: number, attemptId: number, actorToken: string): RuntimeRecord {
  return stateStore.transaction((connection) => {
    const run = stateStore.getRun(rootId, connection);
    const task = stateStore.getTask(taskId, connection);
    const attempt = stateStore.getAttempt(attemptId, connection);
    if (!run || !task || !attempt) throw new RuntimeError("invalid SessionEnd binding");
    if (task.root_id !== rootId || attempt.task_id !== taskId || !stateStore.tokenMatches(actorToken, attempt.actor_token_hash)) {
      throw new RuntimeError("invalid SessionEnd identity");
    }
    stateStore.appendEvent(connection, rootId, "SessionEndObserved", { attempt_state: attempt.state }, taskId, attemptId);
    return { observed: true, attempt_id: attemptId, state: attempt.state };
  });
}

function authorizeRead(rootId: string, actorToken: string): RuntimeRecord {
  stateStore.initializeSchema();
  const run = stateStore.getRun(rootId);
  if (!run) throw new ValueError("run not found");
  if (!stateStore.listAttempts(rootId).some((attempt) => stateStore.tokenMatches(actorToken, attempt.actor_token_hash))) {
    throw new ValueError("invalid actor token");
  }
  return run;
}

export function inspectCurrent(rootId: string, taskId: number, actorToken: string): RuntimeRecord {
  authorizeRead(rootId, actorToken);
  const task = stateStore.getTask(taskId);
  if (!task || task.root_id !== rootId) throw new ValueError("current task does not belong to the authorized run");
  const attempt = stateStore.getCurrentAttempt(taskId);
  if (!attempt || attempt.root_id !== rootId || attempt.task_id !== taskId) throw new ValueError("current attempt binding is invalid");
  const launch = stateStore.getCurrentLaunch(Number(attempt.attempt_id));
  const session = launch ? stateStore.getSessionForLaunch(Number(launch.launch_id)) : null;
  return { run: stateStore.getRun(rootId), task, attempt, launch, session };
}

function identity(): { rootId: string; taskId: number; attemptId: number; actorToken: string } {
  // Validate configuration aliases at the same boundary as the four identity fields.
  compatEnv.value("HOME");
  compatEnv.value("SKILL_DIR");
  const values = compatEnv.validateIdentity();
  const rootId = values.ROOT_ID;
  const actorToken = values.ACTOR_TOKEN;
  if (!rootId || !actorToken || !values.TASK_ID || !values.ATTEMPT_ID) throw new ValueError("missing orchestration identity");
  const taskId = Number(values.TASK_ID);
  const attemptId = Number(values.ATTEMPT_ID);
  if (!Number.isSafeInteger(taskId) || !Number.isSafeInteger(attemptId)) throw new ValueError("orchestration task and attempt IDs must be integers");
  Object.assign(process.env, compatEnv.exportBoth({ ROOT_ID: rootId, TASK_ID: taskId, ATTEMPT_ID: attemptId, ACTOR_TOKEN: actorToken }));
  return { rootId, taskId, attemptId, actorToken };
}

function optionalIdentity(): ReturnType<typeof identity> | null {
  const values = compatEnv.validateIdentity();
  if (!values.ROOT_ID && !values.TASK_ID && !values.ATTEMPT_ID && !values.ACTOR_TOKEN) return null;
  return identity();
}

function hookInput(value: unknown): RuntimeRecord {
  return value !== null && typeof value === "object" && !Array.isArray(value) ? value as RuntimeRecord : {};
}

function skipped(reason: string): RuntimeRecord { return { skipped: true, reason }; }

export function handleHookEvent(kind: string, rawInput: unknown): RuntimeRecord {
  const input = hookInput(rawInput);
  const event = String(input.hook_event_name ?? input.event_name ?? "");
  const expected: Readonly<Record<string, ReadonlySet<string>>> = {
    heartbeat: new Set(["SessionStart", "PostToolUse"]),
    failure: new Set(["PostToolUseFailure"]),
    finish: new Set(["Stop"]),
    clean: new Set(["SessionEnd"]),
  };
  const events = expected[kind];
  if (!events) throw new ValueError("unknown hook event kind");
  if (!events.has(event)) return skipped(`not ${[...events].join(" or ")}`);
  if (kind === "finish" && input.stop_hook_active === true) return {};
  const current = optionalIdentity();
  if (current === null) return skipped("missing orchestration identity");
  if (kind === "failure") {
    return {
      hookSpecificOutput: {
        hookEventName: "PostToolUseFailure",
        additionalContext: "Agents Orchestrator observed a tool failure. Inspect the error before proceeding; retry safely or revise the estimate. Record only reusable pitfalls with write_note, and do not claim completion without the Runtime finish action.",
      },
    };
  }
  if (kind === "heartbeat") {
    if (stateStore.getRun(current.rootId)) hookManager.ensureProjectHooks(process.cwd(), current.rootId);
    return heartbeat(current.rootId, current.taskId, current.attemptId, current.actorToken);
  }
  if (kind === "clean") {
    return observeSessionEnd(current.rootId, current.taskId, current.attemptId, current.actorToken);
  }
  try {
    const inspected = inspectCurrent(current.rootId, current.taskId, current.actorToken);
    const task = inspected.task as RuntimeRecord;
    const attempt = inspected.attempt as RuntimeRecord;
    const terminalTask = new Set(["done", "failed", "blocked", "cancelled"]).has(String(task.status));
    const terminalAttempt = new Set(["done", "failed", "cancelled"]).has(String(attempt.state));
    if (!terminalTask || !terminalAttempt) {
      return {
        decision: "block",
        reason: "Agents Orchestrator Runtime requires this Attempt to submit finish before the Claude session can stop. Submit finish with validation and caveats, or report a failed finish when appropriate.",
      };
    }
  } catch {
    // An unavailable inspection must not create an unbounded Stop loop. Runtime
    // Actions remain authoritative for every lifecycle transition.
  }
  return {};
}

if (import.meta.main) {
  try {
    const command = process.argv[2];
    let result: RuntimeRecord;
    if (command === "hook-event") {
      let input: unknown = {};
      try { input = JSON.parse(await Bun.stdin.text()); } catch { /* missing event is skipped */ }
      result = handleHookEvent(String(process.argv[3] ?? ""), input);
    } else {
      const current = identity();
      if (command === "heartbeat") {
      if (stateStore.getRun(current.rootId)) hookManager.ensureProjectHooks(process.cwd(), current.rootId);
      result = heartbeat(current.rootId, current.taskId, current.attemptId, current.actorToken);
      } else if (command === "inspect-current") result = inspectCurrent(current.rootId, current.taskId, current.actorToken);
      else if (command === "session-end") result = observeSessionEnd(current.rootId, current.taskId, current.attemptId, current.actorToken);
      else throw new ValueError("command must be hook-event, heartbeat, inspect-current, or session-end");
    }
    process.stdout.write(`${JSON.stringify(result)}\n`);
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : "hook runtime failed"}\n`);
    process.exit(2);
  }
}
